### PR TITLE
[triton] bfloat16 aware blocksparse & softmax

### LINKED ISSUE #353: Blocksparse forced to use FP16 when BF16 is supported (for triton v2)
# 🐛 Bug

Within the blocksparse implementation,`q,k,v` are forced to half: https://github.com/facebookresearch/xformers/blob/main/xformers/components/attention/blocksparse.py#L171

After #231 I think blocksparse does support bfloat16 (I tried locally and seems it works). If there are no objections, we should modify the constraint to allow bfloat16. Unsure what the correct approach would be (could use the AMP region to infer the type, or the input type).

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
Fixes #353. Goal was not to touch softmax initially, but it turns out that enabling bfloat16 can die on triton's softmax (tl.max() cannot be lowered in bf16), hence the double fix

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by blefaudeux** [comment]
there's something a bit fishy with the triton softmax and bf16, but other than that seems fine, unit testing is fine locally at least

File: xformers/components/attention/core.py [hunk 1]
```diff
@@ -127,7 +127,7 @@ def _softmax(a: torch.Tensor, causal: bool = False) -> torch.Tensor:
     if a.is_sparse:
         return torch.sparse.softmax(a, dim=a.ndim - 1)
 
-    if _is_triton_available():
+    if _is_triton_available() and a.dtype is not torch.bfloat16:  # FIXME
```
Comment (blefaudeux) [inline xformers/components/attention/core.py:130]: not submitting before this is fixed/tested, this is ugly

File: xformers/triton/softmax.py [hunk 2]
```diff
@@ -181,7 +181,7 @@ def _softmax_dispatch(
     try:
         if torch.cuda.is_available() and x.is_cuda and not _triton_registered_warnings:
             return _softmax_triton.apply(x, mask, log, causal)
-    except (triton.code_gen.OutOfResources, RuntimeError) as e:
```
Comment (blefaudeux) [inline xformers/triton/softmax.py:184]: this exception doesn´t exist anymore

File: xformers/triton/k_softmax.py [hunk 3]
```diff
@@ -158,8 +151,8 @@ def _softmax_backward(
     if causal:
         io_mask = io_mask & (k <= n)
 
-    g = tl.load(grad_out_ptrs, mask=io_mask, other=float(0))
-    o = tl.load(out_ptrs, mask=io_mask, other=float(0))
+    g = tl.load(grad_out_ptrs, mask=io_mask, other=float(0)).to(tl.float32)
```
Comment (blefaudeux) [inline xformers/triton/k_softmax.py:154]: promote to float32 within the kernel, which doesn't change the arithmetics (max and exp were already fp32), and doesn´t touch the bandwidth (cast back to the correct type on the way out, and was of the correct type on the way in). Benchmarks are untouched

File: xformers/components/attention/blocksparse.py [hunk 4]
```diff
@@ -172,10 +172,6 @@ def forward(
                 q.shape[-2], self.block_size
             )
 
-            # Blocksparse only works on fp16
-            q_dtype = q.dtype
-            q, k, v = q.half(), k.half(), v.half()
```
Comment (blefaudeux) [inline xformers/components/attention/blocksparse.py:177]: the whole point :) cc @SeanNaren I didn´t forget you

File: xformers/triton/k_softmax.py [hunk 5]
```diff
@@ -81,17 +79,11 @@ def _softmax(
...
     if use_mask:
         mask_ptrs = M + n * stride_mn + k
-        add_mask = tl.load(mask_ptrs, io_mask, other=float("-inf"))
+        add_mask = tl.load(mask_ptrs, io_mask, other=float("-inf")).to(tl.float32)
         x += add_mask
 
     # compute numerically-stable softmax
     z = x - tl.max(x, axis=0)
-
-    if is_fp16:
```
Comment (blefaudeux) [inline xformers/triton/k_softmax.py:90]: not required anymore, computations inside the kernel are fp32 and that's all fine and dandy

File: tests/test_triton_blocksparse.py [hunk 6]
```diff
@@ -123,17 +120,18 @@ def test_softmax(BLOCK, WIDTH, DTYPE):
     ry = block_sparsify_tensor(ry, layout, BLOCK)
 
     # compare
-    assert_almost_equal(ry, ty)
+    triton.testing.assert_almost_equal(ry, ty)
 
 
 @pytest.mark.skipif(not _triton_available, reason="Triton requires a recent CUDA gpu")
 @pytest.mark.parametrize("block", [32, 43, 128])  # 16, 32,
+@pytest.mark.parametrize("dtype", [torch.float16])
```
Comment (blefaudeux) [inline tests/test_triton_blocksparse.py:128]: parity is not respected with bfloat16, but this amounts to testing triton's blocksparse really so I'm not sure that this is xformers' job really. Lack of parity with pytorch could be explained by different accumulator choices (tf32, float32, ..) and/or normalization in different orders, I tried the obvious suspects to no avail. In the end I thought that it was fair to delegate that to triton, this xformer layer is just a wrapper

File: xformers/triton/k_softmax.py [hunk 7]
```diff
@@ -71,7 +69,7 @@ def _softmax(
     if causal:
         io_mask = io_mask & (k <= n)
 
-    x = tl.load(x_ptrs, mask=io_mask, other=float("-inf"))
+    x = tl.load(x_ptrs, mask=io_mask, other=float("-inf")).to(tl.float32)
```
Comment (blefaudeux) [inline xformers/triton/k_softmax.py:72]: doesn´t affect the IO, this is after the data has been loaded onto the SoC, removes subsequent casts (exp doesn´t accept bfloat16 for instance). Automatically cast back to the relevant type on the way out

**Review by fmassa — APPROVED** [review]
LGTM, thanks a lot @blefaudeux !

**Comment by blefaudeux** [comment]
looks like the CI is not happy for some unrelated reason @fmassa @danthe3rd ? heads up just in case, really sorry if I broke anything but it looked good on the PR

**Comment by danthe3rd** [comment]
Yeah I'm not sure what this is about... I've never seen this:

> 402  Payment Required [IP: 13.56.57.255 443]

Let's ignore it for now
